File: server/app/routers/datasources.py
from __future__ import annotations
"""POST /api/v1/datasources/*"""
from datetime import datetime, timezone

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy import select, or_
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.models import Datasource
from app.utils import safe_int

router = APIRouter(prefix="/api/v1/datasources", tags=["datasources"])

TYPES_WITH_DEPT = {"primarytriage", "leveldepart"}


def _fmt_ds(ds: Datasource) -> dict:
    info = {"id": ds.id, "name": ds.name, "type": ds.type,
            "departmentid": ds.departmentid or [],
            "screenid": ds.screenid,
            "screensplitid": ds.screensplitid or [],
            "queue": ds.queue or [],
            "consultingroomname": ds.consultingroomname or [],
            "windowid": ds.windowid or [],
            "pharmacydeptno": ds.pharmacydeptno,
            "pharmacywinno": ds.pharmacywinno or [],
            "morningcleartime": ds.morningcleartime,
            "afternooncleartime": ds.afternooncleartime}
    if ds.pharmacydeptno and ds.pharmacywinno:
        info["pharmacy"] = f"{ds.pharmacydeptno}_{','.join(str(x) for x in ds.pharmacywinno)}"
    return info


@router.post("/query")
async def query(body: dict, db: AsyncSession = Depends(get_db)):
    stmt = select(Datasource)
    if body.get("name"):
        stmt = stmt.where(Datasource.name.ilike(f"%{body['name']}%"))
    stmt = stmt.order_by(Datasource.type)
    result = await db.execute(stmt)
    return {"errcode": 0, "result": [_fmt_ds(d) for d in result.scalars().all()]}


@router.post("/remove")
async def remove(id: str, db: AsyncSession = Depends(get_db)):
    result = await db.execute(select(Datasource).where(Datasource.id == safe_int(id)))
    ds = result.scalars().first()
    if ds:
        await db.delete(ds)
        await db.commit()
    return {"errcode": 0}


@router.post("/save")
async def save(body: dict, db: AsyncSession = Depends(get_db)):
    now = datetime.now(timezone.utc)

    # 统一解析数组字段：兼容前端发来的数组或字符串
    def _parse_int_list(val) -> list[int]:
        if isinstance(val, list):
            return [int(x) for x in val if str(x).lstrip("-").isdigit()]
        if isinstance(val, str) and val.strip():
            return [int(x.strip()) for x in val.split(",") if x.strip().lstrip("-").isdigit()]
        return []

    def _parse_str_list(val) -> list[str]:
        if isinstance(val, list):
            return [str(x) for x in val if x is not None]
        if isinstance(val, str) and val.strip():
            return [x.strip() for x in val.split(",") if x.strip()]
        return []

    departmentid = _parse_int_list(body.get("departmentid"))
    screensplitid = _parse_int_list(body.get("screensplitid"))
    windowid = _parse_int_list(body.get("windowid"))
    pharmacywinno = _parse_int_list(body.get("pharmacywinno"))
    queue = _parse_str_list(body.get("queue"))
    consultingroomname = _parse_str_list(body.get("consultingroomname")) or None
    ds_type = body.get("type", "")

    # Uniqueness checks — departmentid/JSONB 字段用 Python 列表比对
    if ds_type in TYPES_WITH_DEPT and departmentid:
        r = await db.execute(select(Datasource).where(Datasource.type == ds_type))
        for exist in r.scalars().all():
            if exist.id != safe_int(body.get("id", 0)) and (exist.departmentid or []) == departmentid:
                raise HTTPException(400, "该科室数据源已存在")
    elif ds_type == "secondarytriage" and body.get("screenid"):
        r = await db.execute(select(Datasource).where(Datasource.type == "secondarytriage", Datasource.screenid == body["screenid"]))
        dup = r.scalars().first()
        if dup and dup.id != safe_int(body.get("id", 0)):
            raise HTTPException(400, "该屏ID相关数据源已存在")

    doc_id = body.get("id")
    if doc_id:
        result = await db.execute(select(Datasource).where(Datasource.id == safe_int(doc_id)))
        ds = result.scalars().first()
        if ds:
            for k, v in [("name", body.get("name")), ("type", ds_type), ("departmentid", departmentid),
                         ("screenid", body.get("screenid")), ("screensplitid", screensplitid),
                         ("queue", queue), ("consultingroomname", consultingroomname),
                         ("windowid", windowid), ("pharmacydeptno", body.get("pharmacydeptno")),
                         ("pharmacywinno", pharmacywinno), ("morningcleartime", body.get("morningcleartime")),
                         ("afternooncleartime", body.get("afternooncleartime"))]:
                setattr(ds, k, v)
            ds.ut = now
            await db.commit()
    else:
        ds = Datasource(
            name=body.get("name", ""), type=ds_type, departmentid=departmentid,
            screenid=body.get("screenid"), screensplitid=screensplitid,
            queue=queue, consultingroomname=consultingroomname,
            windowid=windowid, pharmacydeptno=body.get("pharmacydeptno"),
            pharmacywinno=pharmacywinno, morningcleartime=body.get("morningcleartime"),
            afternooncleartime=body.get("afternooncleartime"), ct=now, ut=now,
        )
        db.add(ds)
        await db.commit()
    return {"errcode": 0}

File: smarttriage/lib/controllers/datasources.js
'use strict';

const _ = require('lodash');
const Utils = require('../utils');
const logger = Utils.logger(__filename);
const Model = require('../models');
const Datasource = Model('Datasource');

module.exports = {
  async query(ctx) {
    try {
      const params = ctx.request.body;
      const filter = {};
      if (params.name) {
        filter.name = new RegExp(params.name, 'i');
      }
      const datasources = await Datasource.find(filter).sort({ type: 1 });
      for (let i = 0; i < datasources.length; i++) {
        datasources[i] = datasources[i]._info;
      }
      ctx.body = {
        errcode: 0,
        result: datasources
      };
    } catch (e) {
      logger.error('error', e);
      ctx.status = 500;
      ctx.body = {
        errcode: 10000,
        errmsg: 'server error'
      };
    }
  },
  async remove(ctx) {
    try {
      const params = ctx.request.body;
      await Datasource.remove({ _id: params.id });
      ctx.body = {
        errcode: 0
      };
    } catch (e) {
      logger.error('error', e);
      ctx.status = 500;
      ctx.body = {
        errcode: 10000,
        errmsg: 'server error'
      };
    }
  },
  async save(ctx) {
    try {
      const params = ctx.request.body;
      console.log(params)
      if (!params.datasource) {
        params.datasource = null;
      }
      if (!params.style) {
        params.style = null;
      }
      let datasource;
      const departmentid = [];
      const screensplitid = [];
      const windowid = [];
      const pharmacywinno = [];
      if (params.departmentid) {
        const deptids = params.departmentid.split(',');
        deptids.forEach(function(deptid) {
          if (!isNaN(deptid)) {
            departmentid.push(Number(deptid));
          }
        });
        datasource = await Datasource.findOne({ type: 'primarytriage', departmentid });
        let datasource1 = await Datasource.findOne({ type: 'leveldepart', departmentid });
        if ((datasource && datasource._id != params.id) || (datasource1 && datasource1._id != params.id)) {
          logger.warn('该一级分诊屏ID相关数据源已存在');
          ctx.status = 500;
          ctx.body = {
            errcode: 10001,
            errmsg: '该一级分诊屏ID相关数据源已存在'
          };
          return;
        }
      } else if (params.screenid) {
        datasource = await Datasource.findOne({ type: 'secondarytriage', screenid: params.screenid });
        if (datasource && datasource._id != params.id) {
          logger.warn('该屏ID相关数据源已存在');
          ctx.status = 500;
          ctx.body = {
            errcode: 10001,
            errmsg: '该屏ID相关数据源已存在'
          };
          return;
        }
      } else if (params.screensplitid) {
        const scrsplids = params.screensplitid.split(',');
        scrsplids.forEach(function(scrid) {
          if (!isNaN(scrid)) {
            screensplitid.push(Number(scrid));
          }
        });
        datasource = await Datasource.findOne({ type: 'secondarytriagesplit', screensplitid });
        if (datasource && datasource._id != params.id) {
          logger.warn('该分屏ID相关数据源已存在');
          ctx.status = 500;
          ctx.body = {
            errcode: 10001,
            errmsg: '该分屏ID相关数据源已存在'
          };
          return;
        }
      } else if (params.queue) {
        datasource = await Datasource.findOne({ type: 'secondarytriageultrasonic', queue: params.queue.split(',') });
        if (datasource && datasource._id != params.id) {
          logger.warn('该队列名相关数据源已存在');
          ctx.status = 500;
          ctx.body = {
            errcode: 10001,
            errmsg: '该队列名相关数据源已存在'
          };
          return;
        }
      } else if (params.windowid) {
        const winids = params.windowid.split(',');
        winids.forEach(function(winid) {
          if (!isNaN(winid)) {
            windowid.push(Number(winid));
          }
        });
        datasource = await Datasource.findOne({ type: 'drawbloodtriage', windowid });
        if (datasource && datasource._id != params.id) {
          logger.warn('该窗口编号相关数据源已存在');
          ctx.status = 500;
          ctx.body = {
            errcode: 10001,
            errmsg: '该窗口编号相关数据源已存在'
          };
          return;
        }
      } else if (params.pharmacydeptno || params.pharmacywinno) {
        const pharwinnos = params.pharmacywinno.split(',');
        pharwinnos.forEach(function(winno) {
          if (!isNaN(winno)) {
            pharmacywinno.push(Number(winno));
          }
        });
        datasource = await Datasource.findOne({ type: params.type,
         pharmacydeptno: params.pharmacydeptno, pharmacywinno });
        if (datasource && datasource._id != params.id) {
          logger.warn('该部门编号下该窗口编号相关数据源已存在');
          ctx.status = 500;
          ctx.body = {
            errcode: 10001,
            errmsg: '该部门编号下该窗口编号相关数据源已存在'
          };
          return;
        }
      }
      if (!params.id) {
        datasource = new Datasource({ name: params.name,
          departmentid,
          screenid: params.screenid,
          screensplitid,
          windowid,
          pharmacydeptno: params.pharmacydeptno,
          pharmacywinno,
          type: params.type,
          morningcleartime: params.morningcleartime,
          afternooncleartime: params.afternooncleartime,
          ct: new Date(),
          ut: new Date() });
        if (params.queue) {
          datasource.queue = params.queue.split(',');
          if(params.consultingroomname){
            datasource.consultingroomname=params.consultingroomname.split(',');
            if(datasource.consultingroomname.length>datasource.queue.length){
              logger.warn('诊室名称数量大于队列数量');
              ctx.status = 500;
              ctx.body = {
                errcode: 10001,
                errmsg: '诊室名称数量大于队列数量'
              };
              return;
            }else if(params.consultingroomname.indexOf('，')>=0){

              logger.warn('诊室名称格式错误');
              ctx.status = 500;
              ctx.body = {
                errcode: 10001,
                errmsg: '诊室名称格式错误'
              };
              return;

            }
          }


        }
      } else {
        datasource = await Datasource.findOne({ _id: params.id });
        _.assign(datasource, _.pick(params, ['name', 'type', 'screenid', 'pharmacydeptno', 'morningcleartime', 'afternooncleartime'
      ]), { departmentid, screensplitid, windowid, pharmacywinno });
        if (params.queue) {
          datasource.queue = params.queue.split(',');
          if(params.consultingroomname){
            datasource.consultingroomname=params.consultingroomname.split(',');
            if(datasource.consultingroomname.length > datasource.queue.length){
              logger.warn('诊室名称数量大于队列数量');
              ctx.status = 500;
              ctx.body = {
                errcode: 10001,
                errmsg: '诊室名称数量大于队列数量'
              };
              return;

            }else if(params.consultingroomname.indexOf('，')>=0){

              logger.warn('诊室名称格式有错误');
              ctx.status = 500;
              ctx.body = {
                errcode: 10001,
                errmsg: '诊室名称格式错误'
              };
              return;

            }
          }else{
            datasource.consultingroomname=null;
          }
        }
      }
      await datasource.save();
      ctx.body = {
        errcode: 0
      };
    } catch (e) {
      logger.error('error', e);
      ctx.status = 500;
      ctx.body = {
        errcode: 10000,
        errmsg: 'server error'
      };
    }
  }
};
